### DISCUSSION TITLE
[osu!lazer] Where do skins and beatmaps get stored?

### DISCUSSION THREAD

**Discussion by AlejandroSuero** [ppy/osu#29697] [Q&A] [answered]
When downloading a skin or beatmap and dragging it or double clicking on it, it will load on the game as expected.

But my question comes from wanting to tweak some skins and/or create my own, is there a location where the game will look up or where are they stored?

> [!Note]
> I know that the default skins for **osu!lazer** are drawn on the fly and when exported it only displays in `JSON` https://github.com/ppy/osu/discussions/29525.

I am using a MacOS M1 and when searching for a skin that I downloaded and it is currently being used in the game, it will find nothing.

---

Example:

I've downloaded two different versions of the `WhiteCat` skin, one is already loaded in the game but the other one is not, when searching up `WhiteCat` it only finds one item matching this.

![Screenshot 2024-09-04 at 20 14 00](https://github.com/user-attachments/assets/046b7529-8905-45a1-9377-4f45dce016c2)

> [!note]
> I used the command `find` in the terminal and the finder to get the same result.

---

I know from last year or so when I played **osu!stable** on my Windows machine that the game had a directory in which to create skin directories, and if imported as `.osk` it will be loaded in the game but will not appear in that directory.

How does **osu!lazer** work in that regard? And in my case, is there a change for MacOS?

**Comment by cihe13375** [comment]
afaik all map and skin files in osu!lazer is stored in the `files` subfolder under the main osu!lazer folder (idk what's the default of that on macOS but maybe try searching e.g. `framework.ini` to find it).  All files are named by their hashes (e.g. `00a126dda6ce0811d393a464d04c8b9d3308630e3d62f81ec76dd9aa5d0fb20c`) so they can't be found by their original names. Also they can't be edited directly like in stable.

Beatmap files can be manually edited by using the "edit externally" feature in the editor. This is probably not possible for skins yet but it may come one day I think.

**Reply by AlejandroSuero** [reply]
In osu!lazer is it possible to, for example, create a folder inside `files` with a custom skin?

```txt
└── + osu
    └── + files
        └── + customSkin
            └── cursor.png
```

Or is there a way in lazer? Or is there anything planned to extend the skin editor to be able to tweak or create skins inside the game?

> PS: thank you for telling me about the `framework.ini`, it really helped

**Reply by peppy** [reply, marked as answer]
Will be available soon. You can get a preview of how it will work as external editing is already supported in the beatmap editor (file menu).

**Reply by AlejandroSuero** [reply]
@peppy thanks, I'll be on the lookout for it
